pkg/progress/progress.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package progress
+
+import (
+	"sync"
+	"time"
+)
+
+// Manager is used to maintain the progresses we are care about.
+type Manager struct {
+	sync.RWMutex
+	progesses map[string]*progressIndicator
+}
+
+// NewManager creates a new Manager.
+func NewManager() *Manager {
+	return &Manager{
+		progesses: make(map[string]*progressIndicator),
+	}
+}
+
+// progressIndicator reflects a specified progress.
+type progressIndicator struct {
+	total       float64
+	current     float64
+	startTime   time.Time
+	speedPerSec float64
+}
+
+// Reset resets the progress manager.
+func (m *Manager) Reset() {
+	m.Lock()
+	defer m.Unlock()
+
+	m.progesses = make(map[string]*progressIndicator)
+}
+
+// AddOrUpdateProgress adds a progress into manager if it doesn't exist.
+func (m *Manager) AddOrUpdateProgress(progress string, current float64) (exist bool) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, exist = m.progesses[progress]; exist {
+		m.progesses[progress].current = current
+		if m.progesses[progress].total < current {
```

Comment:
Could you please create a variable pointer to `m.progress[progress]` on L56, so we can use the pointer directly when we want to visit or change the value. Using `m.progress[progress]` to find the pointer in the map is expensive: ref https://github.com/golang/go/blob/master/src/runtime/map.go#L395